crates/dekoder/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+
+use crate::inventory::HeaderT;
+
+/// The EKO errors.
+#[derive(Error, Debug)]
+pub enum EKOError {
+    #[error("No working directory")]
+    NoWorkingDir,
+    #[error("I/O error")]
+    IOError(#[from] std::io::Error),
+    #[error("No target path given")]
+    NoTargetPath,
+    #[error("Target path `{0}` already exists")]
+    TargetAlreadyExists(PathBuf),
+    #[error("Loading operator from `{0}` failed")]
+    OperatorLoadError(PathBuf),
+    #[error("Failed to read key(s) `{0}`")]
+    KeyError(String),
+}
+
+/// My result type has always my errros.
+type Result<T> = std::result::Result<T, EKOError>;
+
+/// A reference point in the evolution atlas.
+pub struct EvolutionPoint {
+    /// Evolution scale.
+    pub scale: f64,
+    /// Number of flavors
+    pub nf: i64,
+}
+
+impl inventory::HeaderT for EvolutionPoint {
+    /// Load from yaml.
+    fn load_from_yaml(yml: &Yaml) -> Result<Self> {
+        // work around float representation
+        let scale = yml["scale"].as_f64();
+        let scale = if scale.is_some() {
+            scale.ok_or(EKOError::KeyError(
+                "because failed to read scale as float".to_owned(),
+            ))?
+        } else {
+            yml["scale"].as_i64().ok_or(EKOError::KeyError(
+                "because failed to read scale as float from int".to_owned(),
+            ))? as f64
+        };
+        let nf = yml["nf"]
+            .as_i64()
+            .ok_or(EKOError::KeyError("because failed to read nf".to_owned()))?;
+        Ok(Self { scale, nf })
+    }
+
+    /// (Protected) comparator.
+    fn eq(&self, other: &Self, ulps: i64) -> bool {
+        self.nf == other.nf && approx_eq!(f64, self.scale, other.scale, ulps = ulps)
+    }
+}
+
+impl EvolutionPoint {
+    /// Comparator.
+    pub fn equals(&self, other: &Self, ulps: i64) -> bool {
+        self.eq(other, ulps)
+    }
+}
+
+/// 4D evolution operator.
+pub struct Operator {
+    pub op: Option<Array4<f64>>,
+}
+
+impl inventory::ValueT for Operator {
+    const FILE_SUFFIX: &'static str = "npz.lz4";
+    fn load_from_path(&mut self, p: PathBuf) -> Result<()> {
+        let mut reader =
+            BufReader::new(FrameDecoder::new(BufReader::new(File::open(p.to_owned())?)));
+        let mut buffer = Vec::new();
+        std::io::copy(&mut reader, &mut buffer)?;
+        let mut npz = NpzReader::new(Cursor::new(buffer))
+            .map_err(|_| EKOError::OperatorLoadError(p.to_owned()))?;
+        let operator: Array4<f64> = npz
+            .by_name("operator.npy")
+            .map_err(|_| EKOError::OperatorLoadError(p.to_owned()))?;
+        self.op = Some(operator);
+        Ok(())
+    }
+}
+
+impl Operator {
+    /// Empty initializer.
+    pub fn new() -> Self {
+        Self { op: None }
+    }
+}
+
+/// EKO output
+pub struct EKO {
+    /// Working directory
+    path: PathBuf,
+    /// Associated archive path
+    tar_path: Option<PathBuf>,
+    /// allow content modifications?
+    read_only: bool,
```

Comment:
This probably mirrors how the implementation choices from Python (?), but I'd drop both `tar_path` and `read_only` and instead always allow changes and require an explicit path when writing out the modified EKO. To me it seems like an unnessarily tight coupling between the EKO object and its representation as a file. What was the reasoning behind this?

Reply:
- done in 8d0d759a
- this was mostly because me/we thinking of an EKO object associated with one thing; and a file is more of "a thing" to me then a directory, but I guess it works just the same
- this basically lifts the file management to the user (from the library) which is just as fine
- the "clean up function" is currently called `destroy` - any suggestions for a better name? just `remove`?

Reply:
but since in the end we will deliver files, also to PineAPPL, we need to have the `extract` also in Rust. As said above it is now simply the users responsibility to keep track of the file

Reply:
OK that makes sense.